Profiler+Kernel: Fix erroneous Loader.so "_end" when profiling Profiler

stendavid opened the pull request:
Profiling Profiler itself gave a large number of "_end" from Loader.so
in the profile instead of the correct symbols.

Profiler mmaps Loader.so to inspect it. It therefore appeared on two
places in memory in the profile, which caused Profiler to believe that
it spanned the whole memory in between and map all adresses to it.

Fix by changing the kernel to map Loader.so as "/usr/bin/Loader.so:
.[section]", and use that in Profiler to skip the second region.

(Loader.so is a special case. Other libraries were already filtered by
skipping mmaps without ':' in the name).

stendavid commented:
<img width="807" alt="perf-with-end" src="https://user-images.githubusercontent.com/5050927/185515152-b6404c08-6c93-4793-9f38-be4b679f3b05.png">

timschumi commented:
This feels a bit like a band-aid fix. Have you looked into whether you can keep it from generating invalid entries at all?

stendavid commented:
Not a lot more hackish than all other double-mapped libraries that are filtered out because they don't have ":" in the name :)

But you're right, there should be a better way. Maybe the offset in the mapped file can be added to the profiling event, then Profiler wouldn't have to try to merge regions at all.

I'll see what I can do.

timschumi commented:
> Not a lot more hackish than all other double-mapped libraries that are filtered out because they don't have ":" in the name.

I believe that's because binaries that are loaded by the dynamic loader have a `DYN_ELF:` prefix or something like that (although we probably would want to test that before making such assumptions).

Maybe we could just make the kernel map `Loader.so` as `DYN_ELF: /usr/lib/Loader.so` instead, and save on the special casing for the Loader?

stendavid commented:
I updated the PR now so that the kernel mmaps Loader.so as `DYN_ELF: /usr/lib/Loader.so` and changed Profiler to use it. Seems to work fine.

I still think would be kind of nice to add the offset of a region in the mapped vmobject to the mmap profiler event, but there's no real need if this is OK :)

timschumi commented:
Whoops, I mixed something up there. `DYN_ELF:` is only used by the loader while initially mapping libraries to get their information. The final mappings that are actually used while the program is running take the form of `/path/to/library: .section` (where `.section` is `.rodata`, `.text`, etc.). That is also the format that the Profiler expects if it sees a map name with a `:` in it.

If we follow that scheme, we can also remove the special case for the loader entirely, as it can just extract the path from the substring between the beginning of the name and the `:`.

If we want to get the section names correct, the following should match what the Dynamic Loader is doing:

- relro region: `.relro` (shouldn't be a thing for the Loader)
- load region:
  - writable region (manually copied): `.data`
  - read-only region (mmapped):
    - executable: `.text`
    - non-executable: `.rodata`

Again, sorry for the confusion. ^^

stendavid commented:
I updated the PR again, now the regions become `/usr/lib/Loader.so: .text`,`/usr/lib/Loader.so: .data`,  etc.
Nice and consistent with the dynamic loader.

timschumi commented on Userland/DevTools/Profiler/Process.cpp:
```diff
@@ -70,10 +70,10 @@ static MappedObject* get_or_create_mapped_object(String const& path)
     StringView path;
+    if (!name.contains(':'))
+        return;
     if (name.contains("Loader.so"sv))
         path = "Loader.so"sv;
```
I'm pretty sure this can be removed completely now, as we can now get the path to the Loader by just parsing the mapping name as usual. That also means that we can also make the following line look like the following and remove the `else` branch completely:

```
auto path = name.substring_view(0, name.view().find(':').value());
```

Also, the second commit deserves a new title and description then. :^)

stendavid replied:
You're right. I've updated the PR

timschumi reviewed (approved):
LGTM!
